libs/community/langchain_community/chat_models/oci_generative_ai.py

```diff
@@ -0,0 +1,363 @@
+
+        extra = Extra.forbid
+
+    @property
+    def _llm_type(self) -> str:
+        """Return type of llm."""
+        return "oci_generative_ai_chat"
+
+    @property
+    def _provider_map(self) -> Mapping[str, Any]:
+        """Get the provider map"""
+        return {
+            "cohere": CohereProvider(),
+            "meta": MetaProvider(),
+        }
+
+    @property
+    def _provider(self) -> Any:
+        """Get the internal provider object"""
+        return self._get_provider(provider_map=self._provider_map)
+
+    def _prepare_request(
+        self,
+        messages: List[BaseMessage],
+        stop: Optional[List[str]],
+        kwargs: Dict[str, Any],
+        stream: bool,
+    ) -> Dict[str, Any]:
+        try:
+            from oci.generative_ai_inference import models
+
+        except ImportError as ex:
+            raise ModuleNotFoundError(
+                "Could not import oci python package. "
+                "Please make sure you have the oci package installed."
+            ) from ex
+        oci_params = self._provider.messages_to_oci_params(messages)
+        oci_params["is_stream"] = stream  # self.is_stream
+        _model_kwargs = self.model_kwargs or {}
+
+        if stop is not None:
+            _model_kwargs[self._provider.stop_sequence_key] = stop
+
+        chat_params = {**_model_kwargs, **kwargs, **oci_params}
+
+        if self.model_id.startswith(CUSTOM_ENDPOINT_PREFIX):
+            serving_mode = models.DedicatedServingMode(endpoint_id=self.model_id)
+        else:
+            serving_mode = models.OnDemandServingMode(model_id=self.model_id)
+
+        request = models.ChatDetails(
+            compartment_id=self.compartment_id,
+            serving_mode=serving_mode,
+            chat_request=self._provider.oci_chat_request(**chat_params),
+        )
+
+        return request
+
+    def _generate(
+        self,
+        messages: List[BaseMessage],
+        stop: Optional[List[str]] = None,
+        run_manager: Optional[CallbackManagerForLLMRun] = None,
+        **kwargs: Any,
+    ) -> ChatResult:
+        """Call out to a OCIGenAI chat model.
+
+        Args:
+            messages: list of LangChain messages
+            stop: Optional list of stop words to use.
+
+        Returns:
+            LangChain ChatResult
+
+        Example:
+            .. code-block:: python
+
+               messages = [
+                            HumanMessage(content="hello!"),
+                            AIMessage(content="Hi there human!"),
+                            HumanMessage(content="Meow!")
+                          ]
+
+               response = llm.invoke(messages)
+        """
+        if self.is_stream:
+            stream_iter = self._stream(
+                messages, stop=stop, run_manager=run_manager, **kwargs
+            )
+            return generate_from_stream(stream_iter)
+
+        request = self._prepare_request(messages, stop, kwargs, stream=False)
```

Comment:
is tool calling supported?

Reply:
tool support will be available soon. we will submit a follow-up PR for this.